libres/lib/include/ert/res_util/memory.hpp

```diff
@@ -0,0 +1,63 @@
+/*
+   Copyright (C) 2011  Equinor ASA, Norway.
+
+   The file 'memory.hpp' is part of ERT - Ensemble based Reservoir Tool.
+
+   ERT is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   ERT is distributed in the hope that it will be useful, but WITHOUT ANY
+   WARRANTY; without even the implied warranty of MERCHANTABILITY or
+   FITNESS FOR A PARTICULAR PURPOSE.
+
+   See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html>
+   for more details.
+*/
+
+#include <ert/python.hpp>
+
+#ifndef ERT_MEMORY_H
+#define ERT_MEMORY_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+long SystemRamFree(void);
+long ProcessMemory(void);
+long ProcessMaxMemory(void);
+long ProcessMaxRSS(void);
+
+class ScopedMemoryLogger {
+public:
+    ScopedMemoryLogger(std::shared_ptr<ert::ILogger> logger,
+                       const std::string message)
+        : m_logger(logger), m_message(message) {
+        m_cur_mem = ProcessMemory();
+        m_max_mem = ProcessMaxMemory();
+
+        m_logger->info("Enter {} Mem: {}Mb  Maxmem: {}Mb", m_message, m_cur_mem,
+                       m_max_mem);
+    };
+
+    ~ScopedMemoryLogger() {
+        long cur_mem = ProcessMemory();
+        long max_mem = ProcessMaxMemory();
+
+        m_logger->info("Exit {} Mem: {}Mb ({}Mb) MaxMem: {}Mb ({}Mb)",
+                       m_message, cur_mem, cur_mem - m_cur_mem, max_mem,
+                       max_mem - m_max_mem);
+    };
+
+private:
+    std::shared_ptr<ert::ILogger> m_logger;
+    std::string m_message;
```

Comment:
Suggestion: I would create a small struct:
```C++
struct meminfo {
    long current; 
    long max;

    meminfo();
}
```
with the implementation of the constructor in the `.cpp` file. The benefits from this:
1. You codify the fact that you are interested in a snapshot of the memory situation - not individual independent values.
2. You avoid exporting the `SystemRamFree(),....` functions (export a struct instead through).
3. You can test the OS call functions by creating `memino` instance - without offering public access into the logger class.

For the member variable in the logger I would avoid the use of `cur` or similar abbreviations for current, when that information is used in the destructor it represents the *initial* memory information.

My previous experience with memory logging like this is that it is difficult to get conclusive results, unless you are doing something really extreme (e.g. spikes of 98% memory usage) the OS will be in there and do it's thing. I think memory reporting by the kernel is a topic which has been worked on - so hopefully you will have more success.

Reply:
Mmmm...  guess we have slightly different views on some aspects here... 😄 

The `SystemRawFree()...` etc are meant to be exposed as utilities. Assuming they produce correct results, they may be useful in developing code which adapts to memory-requirements, not only for logging.

The class utilizes the RAII-promise of C++ to easily "checkpoint" mem-usage and gain insight into how a code-block (scope) uses memory (note logging of max-mem). I.e. we log current and max mem-usage on entering a scope and the same on exiting the scope. Thus, if max-mem changed significally we know the the scope allocated (and hopefully released) a lot of memory, providing insight without the need for analyzing code in detail.

I'll have a look at the prefix of member-vars - names may be clearer, yes...

Finally, I completely agree on the complexity of determining memory usage in general. This is an initlal approach which I believe we can get some useful info from it. It's a shame this will only work in Linux, but I guess that is fair enough for real life.